[JIT] Frozen Graph Linear-BatchNormNd Folding

min-jean-cho opened the pull request:
This PR adds linear-batchnormNd folding for JIT frozen graphs. 

**Performance benchmark**
A preliminary benchmark with a simple model of linear+bn1d tested on first socket, physical cores of skylake machine. 

**FP32, JIT** 
without linear-bn folding
![Screenshot (1368)](https://user-images.githubusercontent.com/93151422/195168944-cfc5b920-bc82-4be1-a221-d194c8fa6c18.png)

with linear-bn folding
![Screenshot (1367)](https://user-images.githubusercontent.com/93151422/195168926-267b0515-45a1-4f08-922d-c150845199ae.png)


cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @mingfeima @XiaobingSuper @ashokei @jingxu10

min-jean-cho commented:
@eellison @davidberard98, could you please help review this PR?

davidberard98 commented:
I started the tests. I will review after they run. If I forget feel free to tag me @min-jean-cho.

Also, please sign the CLA when you get a chance.

davidberard98 commented:
@min-jean-cho I see some build failures. Please take a look. If you aren't sure what the issues are, ping me and I can help investigate.

min-jean-cho commented:
Thank you @davidberard98 for the help, I will ping you after resolving the build issues. Thank you!

min-jean-cho commented:
> Also, please sign the CLA when you get a chance.

I have signed the CLA, seems like it's taking some time to process.

davidberard98 commented:
@min-jean-cho I see this in the CLA which is still failing:

> ❌ The commit (https://github.com/pytorch/pytorch/commit/9c117bf97addfaac5ad5aab70c7080ac66451a34, https://github.com/pytorch/pytorch/commit/1f4087af3a0f820a43136fb0d9bda3d2159cd6d0, https://github.com/pytorch/pytorch/commit/8f3cc17f5ded05b27573a4e4a330d7add2d15818, https://github.com/pytorch/pytorch/commit/4a0db24979fc65af1bbbe55acfde479a067560a4). This user is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

sanchitintel commented:
@min-jean-cho, I think the CLA gets processed immediately with EasyCLA.
It looks like 4 of your commits were pushed with git config that didn't allow EasyCLA to match `user.email` that it has in its database.

If only the latest commit is considered by EasyCLA, then rebasing with GitHub website UI (as you did earlier once in this PR) would help pass this check.
If that wouldn't help, you can update your git config locally, probably soft reset these 5 commits, then create a single commit, and force push to GitHub.

min-jean-cho committed:
added jit frozen grph lin-bn folding

min-jean-cho commented:
Thank you @davidberard98  @sanchitintel  for the guidance. CLA has passed now.

min-jean-cho commented:
Hi @davidberard98, looks like the meta internal CI/CD has passed as well. Could you please help review when available? Thanks!

sanchitintel commented:
@min-jean-cho, that checkmark only suggests that the commit of this PR imported for Meta internal CI checks is the same as the latest commit of this PR, but the status of those CI checks is not visible to folks outside Meta.

davidberard98 reviewed (commented):
@min-jean-cho minor nit: could you use linux newlines instead of dos newlines in the new files you added, just for consistency in the codebase?

still looking at the content of the PR, just wanted to mention that before I forget :)

min-jean-cho commented:
> @min-jean-cho minor nit: could you use linux newlines instead of dos newlines in the new files you added, just for consistency in the codebase?
still looking at the content of the PR, just wanted to mention that before I forget :)

Thanks @davidberard98 :) . I double checked that all files were already in linux newlines. Could you help double check this?

sanchitintel commented:
> minor nit: could you use linux newlines instead of dos newlines in the new files you added, just for consistency in the codebase?

Thanks, this reminded me that I might also need to fix this in the code already merged, and should proactively check for it in future PRs.

sanchitintel commented:
@pytorchbot rebase -b master

sanchitintel commented:
@min-jean-cho, I rebased so as to run the new CI job that was added a few hours ago to detect Windows-style newlines.

min-jean-cho committed:
lint dos to unix

davidberard98 commented on torch/csrc/jit/passes/fold_linear_bn.h:
```diff
@@ -0,0 +1,29 @@
+namespace torch {
+namespace jit {
+
+struct TORCH_API LinearBNParameters {
```
this is a super minor nit - but from what I can tell the contents of fold_linear_bn.(h|cpp) are essentially implementation details of frozen_linear_folding.cpp, is that correct? if so we can probably move this into the anonymous namespace of frozen_linear_folding.cpp right?

(this is mostly just a matter of preference tbh)

min-jean-cho replied:
That's correct. Actually, `fold_linear_bn.h` is mostly copied from `fold_conv_bn.h`, so I thought I would follow it's style. I'm fine with either options -- what do you think?

davidberard98 replied:
I guess it's fine as it is

davidberard98 commented on torch/csrc/jit/passes/frozen_linear_folding.cpp:
```diff
@@ -0,0 +1,119 @@
+
+      // implementation taken from torch/nn/utils/fusion.py
+      Tensor linear_b;
+      if (linear->namedInput("bias")->type() == NoneType::get()) {
```
this looks like it's mostly copied from frozen_conv_folding.cpp, is that accurate? recently we had some issues with autocasting, see https://github.com/pytorch/pytorch/pull/77617. Can you add some tests like this to confirm that it's not an issue in this case?

min-jean-cho replied:
> this looks like it's mostly copied from frozen_conv_folding.cpp, is that accurate?

That's correct, this is mostly copied from frozen_conv_folding.cpp. 

> recently we had some issues with autocasting, see #77617

Thanks for the notice. I verified that similar issue occurs for jit autocasting for lin-bn1d only and on gpu only (inputs are only casted to half on gpu). 

The issue occurs with {2,3}d input tensor, which is the correct usage of bn1d; however, the issue will also occur for an incorrect usage of bn{2,3}d with {2,3}d input tensor. Below, replace `nn.BatchNorm1d(out_features)` with `nn.BatchNorm{2,3}d(out_features)`.

To reproduce `RuntimeError: self and mat2 must have the same dtype`:
```
import torch
import torch.nn as nn

class LinearBN(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(LinearBN, self).__init__()
        self.linear = nn.Linear(in_features, out_features, bias=False)
        self.bn = nn.BatchNorm1d(out_features)
    
    def forward(self, x):
        x = self.linear(x)
        return self.bn(x)

net = LinearBN(32, 32).cuda().eval()  
print(net.linear.weight.dtype) #torch.float32
img = torch.rand((1, 32, 32)).cuda()
with torch.cuda.amp.autocast(True):
    net = torch.jit.script(net)
    net = torch.jit.freeze(net)
print(net.graph)
lin_node = net.graph.findNode("aten::linear", True)
print(lin_node.namedInput("weight").type().dtype()) #torch.float16
print(lin_node.namedInput("bias").type().dtype()) #torch.float32
net(img)
```

min-jean-cho committed:
cast bn1d bias to half if weight half for gpu autocast fold

min-jean-cho committed:
remove debug cmts

min-jean-cho commented on torch/csrc/jit/passes/frozen_linear_folding.cpp:
```diff
@@ -0,0 +1,134 @@
+        at::ScalarType bias_dtype = bn_rm.scalar_type();
+        at::ScalarType weight_dtype = linear_w.scalar_type();
+        at::DeviceType weight_device = linear_w.device().type();
+        if (isBatchNorm1d && weight_device == at::kCUDA &&
```
I'm not sure if we're expected to bypass this issue with bn{2,3}d (assuming correct usage). But for now, I've only checked for bn1d case here.

min-jean-cho commented:
Thanks @davidberard98 for the review and help. I apologize for the very long overdue -- took some time to get access to nv gpu machine.

davidberard98 reviewed (commented):
Thanks for addressing the comments. Regarding BatchNorm1d vs BatchNorm{2,3}d, I think it would be better if we didn't depend on global graph behavior (i.e. whether or not that constant exists in the graph. Would it be okay to always cast to fp16 if for any BatchNorm{k}d to simplify this?

min-jean-cho commented:
> Thanks for addressing the comments. Regarding BatchNorm1d vs BatchNorm{2,3}d, I think it would be better if we didn't depend on global graph behavior (i.e. whether or not that constant exists in the graph. 

I agree, distinguishing between BatchNorm{1,2,3}d via the str constant node is little odd.

> Would it be okay to always cast to fp16 if for any BatchNorm{k}d to simplify this?

Yes, this will simplify the nonetype bias dtype check for linear-BatchNorm{k} folding. My only minor concern is if doing so (casting bias to fp16 as needed) would have a negative impact on the accuracy (since linear-BatchNorm{2,3}d folding were fine with fp16 weight and fp32 bias)

min-jean-cho committed:
cast nonetype bias to half for bnNd

min-jean-cho committed:
remove redundant line

min-jean-cho commented:
> Yes, this will simplify the nonetype bias dtype check for linear-BatchNorm{k} folding. My only minor concern is if doing so (casting bias to fp16 as needed) would have a negative impact on the accuracy (since linear-BatchNorm{2,3}d folding were fine with fp16 weight and fp32 bias)

Hi @davidberard98~ let me know if any update on this PR, thanks! 

Adding @jgong5, if he has any comments on this as well, thanks!

davidberard98 reviewed (approved):
LGTM!

min-jean-cho commented:
@pytorchbot rebase -b master

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 commented:
@min-jean-cho guessing these are just flaky failures, rebased the PR and we can check once those jobs re-run

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot rebase -s

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
fyi, the previous failure was also a flaky test; if this one doesn't pass we can probably force merge.

min-jean-cho commented:
Thanks @davidberard98  ~ sounds good !

min-jean-cho commented:
@davidberard98, I see it's merged! Appreciate your help and review to landing one of my first PRs to PyTorch!

davidberard98 commented:
@min-jean-cho Glad to help - and sorry it took me so long, feel free to ping me more frequently if you need code review in the future, because sometimes I miss the notifications

davidberard98 commented:
@pytorchbot revert -m "possibly causing internal build failures, will revert and investigate later" -c ghfirst

davidberard98 commented:
@min-jean-cho sorry about that, we're seeing some internal failures associated with this PR that I need to investigate before we can try re-landing...

do you think you can re-open a second PR with the same contents? (this will help avoid any issues with syncing between the internal and external version of the PR)
